pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -603,6 +618,214 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream ads.AggregatedDiscove
+			}
+			if pilot.WaitAck {
+				_ = s.waitAck(con)
+			}
+
+		case EndpointType:
+			clusters := discReq.GetResourceNames()
+			// clusters and con.Clusters are all empty, this is not an ack and will do nothing.
+			if len(clusters) == 0 && len(con.Clusters) == 0 {
+				adsLog.Warnf("ADS:EDS: unexpected %v %s %v", peerAddr, con.ConID, discReq.String())
+				continue
+			}
+			if len(clusters) == len(con.Clusters) {
+				sort.Strings(clusters)
+				sort.Strings(con.Clusters)
+
+				// Already got a list of endpoints to watch and it is the same as the request, this is an ack without nonce ?
+				if reflect.DeepEqual(con.Clusters, clusters) {
+					adsLog.Warnf("ADS:EDS: ACK without nonce %s %s (%s) %s %s", peerAddr, con.ConID, con.modelNode.ID, discReq.VersionInfo, discReq.ResponseNonce)
+					continue
+				}
+			}
+
+			for _, cn := range con.Clusters {
+				s.removeEdsCon(cn, con.ConID, con)
+			}
+
+			for _, cn := range clusters {
+				s.addEdsCon(cn, con.ConID, con)
+			}
+
+			con.Clusters = clusters
+			adsLog.Debugf("ADS:EDS: REQ %s %s clusters: %d", peerAddr, con.ConID, len(con.Clusters))
+			err := s.pushEds(s.globalPushContext(), con, true, nil)
+			if err != nil {
+				return err
+			}
+			if pilot.WaitAck {
+				_ = s.waitAck(con)
+			}
+
+		default:
+			adsLog.Warnf("ADS: Unknown watched resources %s", discReq.String())
+		}
+
+	}
+}
+
+func (s *DiscoveryServer) handleAck(con *XdsConnection, discReq *xdsapi.DiscoveryRequest, peerAddr string) {
+	con.mu.Lock()
+	switch discReq.TypeUrl {
+	case ClusterType:
+		con.ClusterNonceAcked = discReq.ResponseNonce
+	case ListenerType:
+		con.ListenerNonceAcked = discReq.ResponseNonce
+	case RouteType:
+		routes := discReq.GetResourceNames()
+		routeNonceSent := con.RouteNonceSent
+		routeVersionInfoSent := con.RouteVersionInfoSent
+		if routeNonceSent != discReq.ResponseNonce {
+			adsLog.Debugf("ADS:RDS: Expired nonce received %s %s (%s), sent %s, received %s",
+				peerAddr, con.ConID, con.modelNode.ID, routeNonceSent, discReq.ResponseNonce)
+			rdsExpiredNonce.Inc()
+		}
+		if discReq.VersionInfo == routeVersionInfoSent {
+			var sortedRoutes []string
+			sort.Strings(routes)
+			sortedRoutes = routes
+			if reflect.DeepEqual(con.Routes, sortedRoutes) {
+				adsLog.Debugf("ADS:RDS: ACK %s %s (%s) %s %s", peerAddr, con.ConID, con.modelNode.ID, discReq.VersionInfo, discReq.ResponseNonce)
+			} else {
+				adsLog.Warnf("ADS:RDS: ACK with different routes %s %s (%s) %s %s", peerAddr, con.ConID, con.modelNode.ID, discReq.VersionInfo, discReq.ResponseNonce)
+			}
+			con.RouteNonceAcked = discReq.ResponseNonce
+		}
+	case EndpointType:
+		con.EndpointNonceAcked = discReq.ResponseNonce
+		if len(edsClusters) != 0 {
+			clusters := discReq.GetResourceNames()
+			con.EndpointPercent = int((float64(len(clusters)) / float64(len(edsClusters))) * float64(100))
+		}
+	}
+	con.mu.Unlock()
+	adsLog.Debugf("ADS:%s: ACK %v %v", discReq.TypeUrl, peerAddr, discReq.String())
+	if pilot.WaitAck {
+		select {
```

Comment:
select {}  is for non-blocking send. If for any reason nobody is reading ( as would happen on the reconnect ), we don't want to block. The '10' size in the channel was the first attempt to avoid it - but safer to use async send.